### PR TITLE
Explicit publish

### PR DESCRIPTION
**Pull request by Jesus89**
Closes https://github.com/CartoDB/cartoframes/issues/1485

- Do not upload GeoDataFrames to CARTO by default (avoiding the creation of random tables and API keys)
- Create API keys by default for private datasets with **a unique name** to reuse already created API keys.
- Add a maps_api_key param to the publish method to allow the user to pass an API key for private datasets.

```
myMap = Map([
    Layer(gdf),
    Layer('public_table', credentials=my_creds),
    Layer('private_table', credentials=my_creds)
])
```

```
>>> myMap.publish('my_map', '1234', credentials=my_creds)
Maps API key "bZN8gW3gugxgpTy7M9KkQg" is used for non public datasets ['private_table']
{...}
```

```
>>> myMap.publish('my_map', '1234', credentials=my_creds, maps_api_key='bZN8gW3gugxgpTy7M9KkQg')
{...}
```

### CODE REVIEW

**Comment by Jesus89** [comment]
Wait for this to be deployed: https://github.com/CartoDB/cartodb/pull/15367

File: cartoframes/data/clients/auth_api_client.py [hunk 1]
```diff
@@ -18,19 +19,27 @@ def __init__(self, credentials=None):
...
+            name = 'cartoframes_{}'.format(create_hash(tables_names))
 
-        api_key = self._api_key_manager.create(
-            name=name,
-            apis=apis,
-            tables=tables)
+        try:
+            api_key = self._api_key_manager.create(name, apis, tables)
+        except Exception as e:
+            if str(e) == 'Validation failed: Name has already been taken':
```
Comment (oleurud) [inline cartoframes/data/clients/auth_api_client.py]: it should raise otherwise

File: cartoframes/viz/kuviz.py [hunk 2]
```diff
@@ -19,28 +15,33 @@
...
+        if not isinstance(maps_api_key, str):
+            maps_api_key = self._create_maps_api_keys(layers)
+
+        self._layers = []
+        for layer in layers:
+            layer_copy = copy.deepcopy(layer)
+
+            # Set the Maps API key if the source is private
+            if not layer_copy.source.is_public() and layer_copy.credentials is not None:
+                layer_copy.credentials['api_key'] = maps_api_key
```
Comment (oleurud) [inline cartoframes/viz/kuviz.py]: does it work with base_url only? I don't remember well now, but we were doing more things in `_add_layers_credentials`

Comment (Jesus89) [reply cartoframes/viz/kuviz.py]: Yes. `_add_layers_credentials` was doing the same as Source.get_credentials does. So the behavior is the same

File: cartoframes/data/clients/auth_api_client.py (hunk 1 above)
Comment (Jesus89) [reply cartoframes/data/clients/auth_api_client.py]: :+1:

**Comment by oleurud** [comment]
I have found a problem. In public publication, the master key is shared (default api key is not used), even if you try to use `maps_api_key='default_public'`

**Comment by Jesus89** [comment]
Thanks @oleurud. I'll take a look

**Comment by Jesus89** [comment]
This is only when using public tables. It's because the Credentials api_key is used by default in local publications, and then the API key is replaced only for private tables.

A solution can be to apply always the maps_api_key to both public and private tables in the publications.

**Comment by oleurud** [comment]
Acceptance :green_apple:

**Pull request merged**